### PR TITLE
Flang: __ARM_ARCH and friends defined for AArch64

### PR DESCRIPTION
**Pull request by pawosm-arm**

### CODE REVIEW

**Comment by sscalpone** [comment]
@vjayathirtha-nv, please review. Thanks!

File: lib/Driver/ToolChains/Flang.cpp [hunk 1]
```diff
@@ -598,17 +598,36 @@ void FlangFrontend::ConstructJob(Compilation &C, const JobAction &JA,
   UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__linux");
   UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__linux__");
   UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__NO_MATH_INLINES");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__LP64__");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__x86_64");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__x86_64__");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__LONG_MAX__=9223372036854775807L");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__SIZE_TYPE__=unsigned long int");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__PTRDIFF_TYPE__=long int");
+  switch (getToolChain().getEffectiveTriple().getArch()) {
```
Comment (vjayathirtha-nv) [inline lib/Driver/ToolChains/Flang.cpp]: The common stuff can be factored out.

File: test/Driver/arch-specific-macros.f [hunk 2]
```diff
@@ -0,0 +1,15 @@
+! RUN: %flang -target aarch64 -v -x f77 -### -c - 2>&1 >/dev/null | FileCheck --check-prefix=AARCH64 %s
```
Comment (vjayathirtha-nv) [inline test/Driver/arch-specific-macros.f]: `-v` not needed

File: test/Driver/arch-specific-macros.f [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+! RUN: %flang -target x86_64 -v -x f77 -### -c - 2>&1 >/dev/null | FileCheck --check-prefix=X86_64 %s
+
+! AARCH64: __aarch64__
+! AARCH64: __ARM_ARCH=8
+! AARCH64: __ARM_ARCH__=8
+! AARCH64-NOT: __x86_64__
```
[the remark this answers is not in the document]
Comment (pawosm-arm) [reply test/Driver/arch-specific-macros.f]: That's not what this test case tests. This test case tests only what this patch introduces: awareness of x86_64 and aarch64 (notice -target option for each flang invocation).

File: test/Driver/arch-specific-macros.f (hunk 2 above)
Comment (pawosm-arm) [reply test/Driver/arch-specific-macros.f]: This test file is useless anyway. It uses `%flang` as it assumes that pull request https://github.com/flang-compiler/clang/pull/20 was pulled into frontend driver. Now as flang moved to release_60 and affected configuration file lit.cfg migrated to lit.cfg.py, referenced pull request won't apply any longer.

**Comment by pawosm-arm** [comment]
Simplified version uploaded.

**Comment by vjayathirtha-nv** [comment]
LGTM.

**Comment by sscalpone** [comment]
@gklimowicz You'll want to update the driver to support OpenPower.

**Pull request merged**